File: src/main/java/org/vors/pairbot/service/UserService.java
package org.vors.pairbot.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;
import org.telegram.telegrambots.meta.api.objects.User;
import org.vors.pairbot.model.Event;
import org.vors.pairbot.model.UserInfo;
import org.vors.pairbot.repository.EventRepository;
import org.vors.pairbot.repository.UserRepository;

import java.time.Duration;
import java.util.Date;
import java.util.List;
import java.util.Optional;

@Transactional
@Component
public class UserService {
    private Logger LOG = LoggerFactory.getLogger(getClass());

    @Autowired
    private UserRepository userRepository;
    @Autowired
    private EventRepository eventRepository;
    @Autowired
    private TimeService timeService;

    public UserInfo createAndSaveUser(User user) {
        return userRepository.save(createUserInfo(user));
    }

    private UserInfo createUserInfo(User user) {
        Integer userId = user.getId();
        String firstName = user.getFirstName();
        String lastName = user.getLastName();

        return createUserInfo(userId, firstName, lastName);
    }

    private UserInfo createUserInfo(Integer userId, String firstName, String lastName) {
        return newUserInfo(userId, firstName, lastName);
    }

    private UserInfo newUserInfo(Integer id, String firstName, String lastName) {
        UserInfo newUser = new UserInfo();
        newUser.setUserId(id);
        newUser.setFirstName(firstName);
        newUser.setLastName(lastName);
        newUser.setCreatedDate(new Date());
        return newUser;
    }

    public UserInfo getExistingUser(Integer userId) {
        return findByUserId(userId).orElseThrow(() -> new IllegalStateException("Existing user not found for ID: " + userId));
    }

    public Optional<UserInfo> findByUserId(Integer userId) {
        return userRepository.findByUserId(userId);
    }

    public List<Event> findUpcomingEvents(Duration upcomingIn, Duration scanPeriod) {
        return eventRepository.findByDateBetweenAndAcceptedTrue(
                timeService.nowPlusDuration(upcomingIn.minus(scanPeriod)),
                timeService.nowPlusDuration(upcomingIn));
    }

}

File: src/main/java/org/vors/pairbot/service/UserService.kt
package org.vors.pairbot.service

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import org.telegram.telegrambots.meta.api.objects.User
import org.vors.pairbot.model.Event
import org.vors.pairbot.model.UserInfo
import org.vors.pairbot.repository.EventRepository
import org.vors.pairbot.repository.UserRepository
import java.time.Duration
import java.util.*

@Transactional
@Component
open class UserService(
        @Autowired
        open var userRepository: UserRepository,
        @Autowired
        open var eventRepository: EventRepository,
        @Autowired
        open var timeService: TimeService
) {
    private val LOG = LoggerFactory.getLogger(javaClass)


    fun createAndSaveUser(user: User): UserInfo {
        return userRepository.save(createUserInfo(user))
    }

    private fun createUserInfo(user: User): UserInfo {
        val userId = user.id
        val firstName = user.firstName ?: "null"
        val lastName = user.lastName ?: "null"

        return createUserInfo(userId, firstName, lastName)
    }

    private fun createUserInfo(userId: Long, firstName: String, lastName: String): UserInfo {
        return newUserInfo(userId, firstName, lastName)
    }

    private fun newUserInfo(id: Long, firstName: String, lastName: String): UserInfo {
        return UserInfo(id, firstName, lastName, createdDate = Date())
    }

    fun getExistingUser(userId: Long?): UserInfo {
        return findByUserId(userId).orElseThrow { IllegalStateException("Existing user not found for ID: " + userId!!) }
    }

    fun findByUserId(userId: Long?): Optional<UserInfo> {
        return userRepository.findByUserId(userId)
    }

    fun findByNoEventsAfter(date: Date): List<UserInfo> {
        return userRepository.findByNoEventsAfter(date)
    }

    fun findUpcomingEvents(upcomingIn: Duration, scanPeriod: Duration): List<Event> {
        return eventRepository.findByDateBetween(
                timeService.nowPlusDuration(upcomingIn.minus(scanPeriod)),
                timeService.nowPlusDuration(upcomingIn))
    }

}
